app/uk/gov/hmrc/agentmapping/controller/MappingController.scala

```diff
@@ -21,76 +21,124 @@ import javax.inject.{Inject, Singleton}
-import play.api.mvc.{Action, AnyContent}
+import play.api.mvc.{Action, AnyContent, Request}
 import reactivemongo.core.errors.DatabaseException
 import uk.gov.hmrc.agentmapping.audit.AuditService
 import uk.gov.hmrc.agentmapping.connector.DesConnector
-import uk.gov.hmrc.agentmapping.repository.{Mapping, MappingRepository}
+import uk.gov.hmrc.agentmapping.model.{Identifier, Identifiers, SaAgentReferenceMappings, VatAgentReferenceMappings}
+import uk.gov.hmrc.agentmapping.repository._
 import uk.gov.hmrc.agentmtdidentifiers.model.{Arn, Utr}
 import uk.gov.hmrc.auth.core.AuthProvider.GovernmentGateway
 import uk.gov.hmrc.auth.core._
+import uk.gov.hmrc.auth.core.authorise.Predicate
 import uk.gov.hmrc.auth.core.retrieve.{GGCredId, Retrievals}
-import uk.gov.hmrc.domain.SaAgentReference
+import uk.gov.hmrc.http.HeaderCarrier
 import uk.gov.hmrc.mongo.json.ReactiveMongoFormats
 import uk.gov.hmrc.play.HeaderCarrierConverter.fromHeadersAndSession
 import uk.gov.hmrc.play.microservice.controller.BaseController
 
 import scala.concurrent.Future
 
+class UnsupportedIdentifierKey(identifier: Identifier) extends Exception(s"Unsupported identifier key ${identifier.key}")
+
 @Singleton
-class MappingController @Inject()(mappingRepository: MappingRepository, desConnector: DesConnector, auditService: AuditService, val authConnector: AuthConnector)
+class MappingController @Inject()(vatAgentReferenceMappingRepository: VatAgentReferenceMappingRepository,
+                                  saAgentReferenceMappingRepository: SaAgentReferenceMappingRepository,
+                                  desConnector: DesConnector,
+                                  auditService: AuditService,
+                                  val authConnector: AuthConnector)
   extends BaseController with AuthorisedFunctions {
 
   import auditService._
   import uk.gov.hmrc.play.http.logging.MdcLoggingExecutionContext._
 
-  def createMapping(utr: Utr, arn: Arn, saAgentReference: SaAgentReference): Action[AnyContent] = Action.async { implicit request =>
+  private val validEnrolmentIdentifierKeys: Map[String, String] = Map(
+    "IRAgentReference" -> "IR-SA-AGENT",
+    "VATRegNo" -> "HMCE-VATDEC-ORG"
+  )
+
+  def enrolmentsFor(identifiers: Identifiers): Predicate = {
+    identifiers.get
+      .map(i => Enrolment(
+        validEnrolmentIdentifierKeys.getOrElse(i.key, throw new UnsupportedIdentifierKey(i))
+      ).withIdentifier(i.key, i.value)
+      ).reduce[Predicate](_ or _)
+  }
+
+  def createMapping(utr: Utr, arn: Arn, identifiers: Identifiers): Action[AnyContent] = Action.async { implicit request =>
     implicit val hc = fromHeadersAndSession(request.headers, None)
-    authorised(AuthProviders(GovernmentGateway) and AffinityGroup.Agent and Enrolment("IR-SA-AGENT").withIdentifier("IRAgentReference", saAgentReference.value))
+    authorised(AuthProviders(GovernmentGateway) and AffinityGroup.Agent and enrolmentsFor(identifiers))
```

Comment:
If there are multiple identifiers, `enrolmentsFor(identifiers)` will returns a Predicate that only requires at least 1 of the enrolments, not all of them (in `enrolmentsFor` the enrolments are reduced by ORing rather than ANDing). 
The code below on line 76 looks like it could then go ahead and create a mapping for each of the identifiers even if the user doesn't have the corresponding enrolment.

Reply:
good spot, I blindly copy-pasted it from mapping-frontend

Reply:
It has been fixed, pls look again!